pkg/configs/db/postgres/postgres.go

```diff
@@ -144,16 +145,27 @@ func (d DB) GetRulesConfig(userID string) (configs.VersionedRulesConfig, error)
 }
 
 // SetRulesConfig sets the current alertmanager config for a user.
-func (d DB) SetRulesConfig(userID string, config configs.RulesConfig) error {
-	current, err := d.GetConfig(userID)
-	if err != nil {
-		return err
-	}
-	new := configs.Config{
-		AlertmanagerConfig: current.Config.AlertmanagerConfig,
-		RulesFiles:         config,
-	}
-	return d.SetConfig(userID, new)
+func (d DB) SetRulesConfig(userID string, oldConfig, newConfig configs.RulesConfig) (bool, error) {
+	updated := false
+	err := d.Transaction(func(tx DB) error {
+		current, err := d.GetConfig(userID)
+		if err != nil && err != sql.ErrNoRows {
+			return err
+		}
+		// The supplied oldConfig must match the current config. If no config
+		// exists, then oldConfig must be nil. Otherwise, it must exactly
+		// equal the existing config.
+		if !((err == sql.ErrNoRows && oldConfig == nil) || reflect.DeepEqual(current.Config.RulesFiles, oldConfig)) {
```

Comment:
Less repetition if you moved the details of the check into a `RulesConfig.Equal()` method, and that could be coded as a straightforward loop rather than using `reflect`.

Reply:
FWIW, in my early morning fog, it wasn't that straightforward. I have become too used to just writing `deriving Eq`.